### PR TITLE
feat(reports): Include transactions in project series histogram

### PR DESCRIPTION
**Pull request by PixelDust22**
Supersedes #26025

This PR focuses on adding transaction related data into the weekly report context. This PR should not modify the current email experiences because it still uses the old html template. All changes to the template are done in a separate template, new.html. The actual rollout will be addressed in a different PR.

image
image

### CODE REVIEW

File: src/sentry/tasks/reports.py [hunk 1]
```diff
@@ -917,15 +1049,110 @@ def summarize(key, points):
...
+
+    query = Query(
+        dataset=Dataset.Transactions.value,
+        match=Entity("transactions"),
+        select=[
+            Column("transaction_name"),
+            Function("quantile(0.95)", [Column("duration")], "q95"),
+        ],
+        where=[
+            Condition(Column("finish_ts"), Op.GTE, start - timedelta(days=7)),
```
Comment (Neo-Zhixing) [inline src/sentry/tasks/reports.py:1112]: Is there a way to merge these two queries?

Comment (vladkluev) [reply src/sentry/tasks/reports.py:1112]: yes, you can query the events for the entire interval that you want and then split the two intervals in python

Comment (Neo-Zhixing) [reply src/sentry/tasks/reports.py:1112]: I just need the p95 of the dataset for both intervals - how would you group by "this week" versus "last week"?

File: src/sentry/templates/sentry/emails/reports/body.html [hunk 2]
```diff
@@ -363,7 +363,7 @@ <h1>
 
     {% with series=report.projects.series height=110 %}
 
-    <h4>Errors by Project</h4>
+    <h4>Events by Project</h4>
```
Comment (scefali) [inline src/sentry/templates/sentry/emails/reports/body.html:366]: @Neo-Zhixing if we aren't including transactions in this yet, I think we should just leave the title as errors

File: src/sentry/tasks/reports.py (hunk 1 above)
Comment (vladkluev) [reply src/sentry/tasks/reports.py:1112]: I think this may be possible but I'll need some time to see if it is. Here is how I imagine it can be done:

Construct a query that gets all the datapoints for the two weeks, e.g.

```python3
time_range_query = Query(
    dataset=Dataset.Transactions.value,
    match=Entity("transactions"),
    select=[
        Column("transaction_name"),
        Column("timestamp"),
        Column("duration"),
    ],
    where=[
        Condition(Column("finish_ts"), Op.GTE, start - timedelta(weeks=2)),
        Condition(Column("finish_ts"), Op.LT, start),
        Condition(Column("transaction_name"), Op.IN, transaction_names),
        Condition(Column("project_id"), Op.IN, project_ids),
    ]
)
```

Use that as a subquery to query for your quantiles

```python3
quantile_queries = Query(
    dataset=Dataset.Transactions.value,
    match=time_range_query,
    ```
    
what exactly that query looks like I'll take some more time to think about.

File: src/sentry/tasks/reports.py [hunk 3]
```diff
@@ -230,25 +234,75 @@ def build_project_series(start__stop, project):
...
+            Condition(Column("project_id"), Op.EQ, project.id),
+            Condition(Column("org_id"), Op.EQ, project.organization_id),
+            Condition(Column("outcome"), Op.EQ, Outcome.ACCEPTED),
+            Condition(
+                Column("category"),
+                Op.IN,
+                [*DataCategory.error_categories(), DataCategory.TRANSACTION],
+            ),
+        ],
+        groupby=[Column("time"), Column("category")],
```
Comment (vladkluev) [inline src/sentry/tasks/reports.py:276]: how many categories can there be?

File: src/sentry/tasks/reports.py (hunk 1 above)
Comment (vladkluev) [reply src/sentry/tasks/reports.py:1112]: Ok it is possible to do this. You can change your quantile to

`quantileIf(95.0)(duration, timestamp > {something}) AS q95LastWeek` 

However it will most likely not yield a performance improvement so two queries is probably fine here. Although since they only differ by time clause I would consider refactoring so there's less duplication between the queries

File: src/sentry/tasks/reports.py [hunk 4]
```diff
@@ -230,25 +234,75 @@ def build_project_series(start__stop, project):
...
+        granularity=Granularity(rollup),
+        orderby=[OrderBy(Column("time"), Direction.ASC)],
     )
+    outcome_series = raw_snql_query(outcomes_query, referrer="reports.series")
```
Comment (evanh) [inline src/sentry/tasks/reports.py:280]: Nit: It would be nice if this had a more descriptive/verbose referrer.

File: src/sentry/tasks/reports.py [hunk 5]
```diff
@@ -438,6 +492,94 @@ def build_project_calendar_series(interval, project):
     return clean_calendar_data(project, series, start, stop, rollup)
 
 
+def build_key_events(interval, project):
+    start, stop = interval
+
+    # Take the 3 most frequently occuring events
```
Comment (evanh) [inline src/sentry/tasks/reports.py:498]: When you say "events" here you are referring to errors, right?

File: src/sentry/tasks/reports.py [hunk 6]
```diff
@@ -781,7 +931,7 @@ def deliver_organization_user_report(timestamp, duration, organization_id, user_
 
     inclusion_predicates = [
         lambda interval, project__report: project__report[1] is not None,
-        has_valid_aggregates,
+        # has_valid_aggregates,
```
Comment (evanh) [inline src/sentry/tasks/reports.py:934]: Why did you remove this?

File: src/sentry/tasks/reports.py (hunk 5 above)
Comment (Neo-Zhixing) [reply src/sentry/tasks/reports.py:498]: yeah lol. changed the name to `build_key_errors`

File: src/sentry/tasks/reports.py (hunk 4 above)
Comment (Neo-Zhixing) [reply src/sentry/tasks/reports.py]: changed the referrer name & added a few for other queries as well.

File: src/sentry/tasks/reports.py (hunk 6 above)
Comment (Neo-Zhixing) [reply src/sentry/tasks/reports.py]: that was in fact accidental, my bad

**Review by evanh — COMMENTED** [review]
The SnQL part looks good to me. I would get approval from someone who understands the rest though.

File: src/sentry/web/frontend/debug/mail.py [hunk 7]
```diff
@@ -512,15 +519,22 @@ def build_report(project):
...
+            key_events=[(g.id, random.randint(0, 1000)) for g in Group.objects.all()[:3]],
+            key_transactions=[("/transaction/1", 1234, project.id, 1111, 2222)],
         )
 
     if random.random() < 0.85:
         personal = {"resolved": random.randint(0, 100), "users": int(random.paretovariate(0.2))}
     else:
         personal = {"resolved": 0, "users": 0}
 
+    if request.GET.get("new"):
```
Comment (scefali) [inline src/sentry/web/frontend/debug/mail.py:531]: @Neo-Zhixing how come we aren't using a feature flag for this?

**Review by scefali — APPROVED** [review]
